library/spdm_common_lib/libspdm_com_msg_log.c

```diff
@@ -0,0 +1,100 @@
+void libspdm_init_msg_log (void *context, void *msg_buffer, size_t msg_buffer_size)
+{
+    libspdm_context_t *spdm_context;
+
+    LIBSPDM_ASSERT((context != NULL) && (msg_buffer != NULL));
+    LIBSPDM_ASSERT(msg_buffer_size != 0);
+
+    spdm_context = context;
+    spdm_context->msg_log.buffer = msg_buffer;
+    spdm_context->msg_log.max_buffer_size = msg_buffer_size;
+    spdm_context->msg_log.buffer_size = 0;
+    spdm_context->msg_log.mode = 0;
+    spdm_context->msg_log.status = 0;
+}
+
+void libspdm_set_msg_log_mode (void *context, uint32_t mode)
+{
+    libspdm_context_t *spdm_context;
+
+    LIBSPDM_ASSERT(context != NULL);
+
+    spdm_context = context;
+    spdm_context->msg_log.mode = mode;
+}
+
+uint32_t libspdm_get_msg_log_status (void *context)
+{
+    libspdm_context_t *spdm_context;
+
+    LIBSPDM_ASSERT(context != NULL);
+
+    spdm_context = context;
+
+    return spdm_context->msg_log.status;
+}
+
+size_t libspdm_get_msg_log_size (void *context)
+{
+    libspdm_context_t *spdm_context;
+
+    LIBSPDM_ASSERT(context != NULL);
+
+    spdm_context = context;
+
+    return spdm_context->msg_log.buffer_size;
+}
+
+void libspdm_reset_msg_log (void *context)
+{
+    libspdm_context_t *spdm_context;
+
+    LIBSPDM_ASSERT(context != NULL);
+
+    spdm_context = context;
+
+    spdm_context->msg_log.buffer_size = 0;
+    spdm_context->msg_log.mode = 0;
+    spdm_context->msg_log.status = 0;
+}
+
+void libspdm_append_msg_log(void *context, void *message, size_t message_size)
+{
+    LIBSPDM_ASSERT((context != NULL) && (message != NULL));
+
+    libspdm_context_t *spdm_context;
+
+    spdm_context = context;
+
+    if (((spdm_context->msg_log.mode & LIBSPDM_MSG_LOG_MODE_ENABLE) != 0) &&
+        ((spdm_context->msg_log.status & LIBSPDM_MSG_LOG_STATUS_CLAMP) == 0)) {
+        if (spdm_context->msg_log.buffer_size + message_size >
+            spdm_context->msg_log.max_buffer_size) {
+            libspdm_copy_mem((uint8_t *)spdm_context->msg_log.buffer +
+               spdm_context->msg_log.buffer_size,
+               spdm_context->msg_log.max_buffer_size, message,
+               spdm_context->msg_log.max_buffer_size - spdm_context->msg_log.buffer_size);
+            spdm_context->msg_log.status |= LIBSPDM_MSG_LOG_STATUS_CLAMP;
+        } else {
+            libspdm_copy_mem((uint8_t *)spdm_context->msg_log.buffer +
+                             spdm_context->msg_log.buffer_size,
+                             spdm_context->msg_log.max_buffer_size, message, message_size);
+            spdm_context->msg_log.buffer_size += message_size;
+        }
+
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "Message Logging Status = [%x] Buffer Size = [%x] "
+                      "Max Buffer Size = [%x]\n", spdm_context->msg_log.status,
+                      spdm_context->msg_log.buffer_size, spdm_context->msg_log.max_buffer_size));
+        libspdm_internal_dump_hex(spdm_context->msg_log.buffer, spdm_context->msg_log.buffer_size);
```

Comment:
Apparently dumping the message log is causing test failures only on Windows (even GCC / Mingw). @jyao1, I can leave this debug info out, but off the top of your head might you know why this would only fail on Windows?

Reply:
I am surprised on that. Does LIBSPDM_DEBUG() cause failure? or does libspdm_internal_dump_hex() cause failure?

Reply:
`libspdm_internal_dump_hex` prints out the entire message log, which gets pretty big, so it might be memory corruption.

Reply:
Will leave this unresolved for now.